### PR TITLE
Construct Variants from Reference properly in GDNative

### PR DESCRIPTION
**Pull request by endragor**

### CODE REVIEW

**Review by karroffel — APPROVED** [review]

**Comment by hpvb** [comment]
Please don't merge this before some of the questions are answered. Particularly the use of dynamic_cast can't be done that way.

**Comment by karroffel** [comment]
Oh yes. Can't he use `Object::cast_to<Reference>`?

**Comment by hpvb** [comment]
Yeah, ideally Object::cast_to<> otherwise a static_cast<> as we don't allow dynamic_cast<> anywhere for targets without RTTI, currently Android and HTML5 (although I hope to build Android with RTTI at some point)

**Review by hpvb — CHANGES_REQUESTED** [review]

File: core/reference.cpp
Comment (karroffel) [reply core/reference.cpp]: This is needed so any script that wants to use the functionality gets notified about refcount changes. This is particularly useful for GDNative when used in languages with integrated GC.

Comment (hpvb) [reply core/reference.cpp]: OK, wouldn't it make more sense to make void Reference::reference() a return a bool, like SafeRefCount::ref(), then call that instead of duplicating the reference() functionality here?

File: core/variant.cpp
Comment (endragor) [reply core/variant.cpp]: It's simply to have less calls to reference/unreference, which use atomic operations and get propagated to the script system. We just need to get the raw pointer and it can be done very cheaply.

File: modules/gdnative/gdnative/variant.cpp
Comment (endragor) [reply modules/gdnative/gdnative/variant.cpp]: Thank you, I was not aware of that.

File: core/reference.cpp
Comment (endragor) [reply core/reference.cpp]: Changed init_ref to use reference()/unreference().

File: modules/gdnative/gdnative/variant.cpp
Comment (hpvb) [inline modules/gdnative/gdnative/variant.cpp]: Can we really continue if cast_to() returns NULL here?

**Review by hpvb — CHANGES_REQUESTED** [review]

File: modules/gdnative/gdnative/variant.cpp
Comment (endragor) [reply modules/gdnative/gdnative/variant.cpp]: Yes, it can be a non-Reference Object and it's fine.

**Review by hpvb — APPROVED** [review]

**Comment by hpvb** [comment]
Thanks for the clarification and fixes.

**Comment by hpvb** [comment]
I'd just like to ask you to amend the commit message to explain the reasons for your change exactly, or rather what was broken previously. This will help a lot if a bisect points to this commit.

**Comment by endragor** [comment]
Done.

**Comment by hpvb** [comment]
OK, will merge once Travis is happy.

**Pull request merged**
